homework 5 threading

alex-zdns opened the pull request.

alex-zdns committed:
Preparation steps

alex-zdns committed:
develop load from JSON movie data

alex-zdns committed:
add dynamic size for movies and actors cards

alex-zdns committed:
fix: close file after text reading (in JsonLoad.kt fun readAssetFileToString())

alex-zdns committed:
line break after ": " in movie title in MovieDetailsFragment.kt

an-yurii commented on app/src/main/java/ru/alexzdns/fundamentals/homework/ui/adapters/ActorsAdapter.kt:
```diff
@@ -14,15 +17,19 @@ class ActorsAdapter(
+        val margin = parent.resources.getDimensionPixelSize(R.dimen.md_margin_side)
+        val side = (Resources.getSystem().displayMetrics.widthPixels - margin * 2) / parent.resources.getInteger(R.integer.actors_list_item_count)
+        holder.avatar.layoutParams.width = side
+        holder.avatar.layoutParams.height = side
```
Параметры не нужно менять динамически. Достаточно зафиксировать их в разметке.

an-yurii commented on app/src/main/java/ru/alexzdns/fundamentals/homework/ui/adapters/ActorsAdapter.kt:
```diff
@@ -14,15 +17,19 @@ class ActorsAdapter(
     var actors: List<Actor>
```
Старайся объявлять иммутабельные переменные
```suggestion
    private val actors: List<Actor>
```

an-yurii commented on app/src/main/java/ru/alexzdns/fundamentals/homework/ui/adapters/MovieAdapter.kt:
```diff
@@ -15,50 +19,68 @@ class MovieAdapter(
+        val itemCount = parent.resources.getInteger(R.integer.movies_list_span_count)
+        val side = (Resources.getSystem().displayMetrics.widthPixels - margin * 2 - marginBetweenCards * (itemCount - 1)) / itemCount
+
+        holder.cardView.layoutParams.width = side
```
Параметры изображений нужно зафиксировать. верстке. Нет нужды их настраивать динамически.

alex-zdns replied:
Мне кажется, что если настраивать размеры динамически, то списки будут на разных устройствах смотреться лучше. Нельзя ли это оставить как доп. фичу? Или сделано слишком криво?

an-yurii replied:
Для этих целей как раз и придуманы dp и разные квалификаторы: hdpi, xhdpi и т.п. Большинство телефонов в портреной ориентации будут иметь ширину 320dp https://developer.android.com/training/multiscreen/screensizes#TaskUseSWQuali

an-yurii commented on app/src/main/java/ru/alexzdns/fundamentals/homework/ui/adapters/MovieAdapter.kt:
```diff
@@ -15,50 +19,68 @@ class MovieAdapter(
+            val imageOption = RequestOptions()
+                .placeholder(R.drawable.mlf_poster_holder)
+                .fallback(R.drawable.mlf_poster_holder)
+                .centerCrop()
```
Нужно сделать статическим объектом, чтобы не пересоздавать при каждом биндинге.
Перенести в `companion object {}`

an-yurii commented on app/src/main/java/ru/alexzdns/fundamentals/homework/ui/fragments/MovieDetailsFragment.kt:
```diff
@@ -46,13 +42,42 @@ class MovieDetailsFragment : androidx.fragment.app.Fragment() {
+
+            findViewById<TextView>(R.id.mdf_tv_movie_title).run {
+                var movieTitle = movie.title
+                if (movieTitle.contains(": ")) {
+                    movieTitle = movieTitle.replace(": ", ":\n")
+                }
```
В манипуляции с заголовком нет необходимости. Пусть строка переносится по словам как придется) В контенте должны передаваться неразрывные пробелы там, где это необходимо.

Если нужна какая-либо модификация данных модели, то правильнее это делать при маппинге данных (например, в методе `parseMovies`). Минимум логики в активити и фрагментах.

an-yurii commented on app/src/main/java/ru/alexzdns/fundamentals/homework/ui/fragments/MovieDetailsFragment.kt:
```diff
@@ -46,13 +42,42 @@ class MovieDetailsFragment : androidx.fragment.app.Fragment() {
+                .into(backdrop)
+
+            findViewById<TextView>(R.id.mdf_tv_movie_title).run {
+                var movieTitle = movie.title
```
Старайся использовать val. В котлин if (try, when, ...) возвращает результат, можно писать конструкции вида
```
val movieTitle = if (movieTitle.contains(": ")) {
        movieTitle = movieTitle.replace(": ", ":\n")
    } else {
        movie.title
    }
```

an-yurii commented on app/src/main/java/ru/alexzdns/fundamentals/homework/ui/fragments/MovieListFragment.kt:
```diff
@@ -6,14 +6,16 @@ import android.view.LayoutInflater
 class MovieListFragment : androidx.fragment.app.Fragment() {
     private var listenerMovieList: MovieListClickListener? = null
     private var recycler: RecyclerView? = null
+    private val scope = CoroutineScope(Dispatchers.IO)
```
Стоило бы добавить обработку исключений и поменять job на SupervisorJob

an-yurii commented on app/src/main/res/layout/view_holder_actor.xml:
```diff
@@ -5,19 +5,29 @@
-        android:contentDescription="@string/actor_photo"
-        app:layout_constraintDimensionRatio="1:1"
-        android:scaleType="centerCrop"
+    <androidx.cardview.widget.CardView
```
Ты что-то переусложняешь верстку. Непонятен замысел с CardView

alex-zdns replied:
С помощью CardView делал закругленные углы у фото актеров (как в фигме). Сейчас удалил и сделал через glade.

alex-zdns committed:
fix view holder adapter

delete cardView from view_holder_actor.xml
add cornersRadius in ActorsAdapter.kt

alex-zdns committed:
fix after code review

alex-zdns committed:
fix actor avatars and card movie sizes
